crates/wasm-sdk/src/host_calls.rs

```diff
@@ -0,0 +1,508 @@
+pub use serde_json::{Map, Value};
+
+/// An HTTP request combines a head consisting of a [`Method`], [`Uri`], and headers with a [`BodyChunk`], which provides
+/// access to the first chunk of a request body.
+pub type Request = http::Request<BodyChunk>;
+/// An HTTP response combines a head consisting of a [`StatusCode`] and headers with a [`BodyChunk`], which provides
+/// access to the first chunk of a response body.
+pub type Response = http::Response<BodyChunk>;
+
+// NOTE: fields are documented via Markdown instead of normal rustdoc because the underlying type is from the macro.
+/// A `Breaker` contains the values needed to implement a circuit-breaker pattern within a plugin.
+///
+/// # Fields
+///
+/// * `generation` - The number of times a breaker has been incremented within the expiration window.
+/// * `successes` - The number of total success outcomes tracked within the expiration window.
+/// * `failures` - The number of total failure outcomes tracked within the expiration window.
+/// * `consecutive_successes` - The number of consecutive success outcomes.
+/// * `consecutive_failures` - The number of consecutive failure outcomes.
+/// * `expiration` - The expiration timestamp in seconds since the epoch.
+pub type Breaker = crate::bulwark_host::BreakerInterface;
+/// A `Rate` contains the values needed to implement a rate-limiter pattern within a plugin.
+///
+/// # Fields
+///
+/// * `attempts` - The number of attempts made within the expiration window.
+/// * `expiration` - The expiration timestamp in seconds since the epoch.
+pub type Rate = crate::bulwark_host::RateInterface;
+
+/// The number of successes or failures to increment the breaker by.
+pub enum BreakerDelta {
+    Success(i64),
+    Failure(i64),
+}
+
+/// The first chunk of an HTTP body.
+///
+/// Bulwark does not send the entire body to the guest plugin environment. This limitation limits the impact of
+/// copying a large number of bytes from the host into guest VMs. A full body copy would be required for each
+/// plugin for every request or response otherwise.
+///
+/// This has consequences for any plugin that wants to parse the body it receives. Some data formats like JSON
+/// may be significantly more difficult to work with if only partially received, and streaming parsers which may be
+/// more tolerant to trunctation are recommended in such cases. There will be some situations where this limitation
+/// prevents useful parsing entirely and plugins may need to make use of the `unknown` result value to express this.
+pub struct BodyChunk {
+    pub end_of_stream: bool,
+    pub size: u64,
+    pub start: u64,
+    // TODO: use bytes crate to avoid copies
+    pub content: Vec<u8>,
+}
+
+/// An empty HTTP body
+pub const NO_BODY: BodyChunk = BodyChunk {
+    end_of_stream: true,
+    size: 0,
+    start: 0,
+    content: vec![],
+};
+
+// NOTE: `pub use` pattern would be better than inlined functions, but apparently that can't be rustdoc'd?
```

Comment:
When I tried to put a `rustdoc` comment over a `pub use`, it simply didn't document the function. Presumably it tried to load the `rustdoc` from the original function declaration, which doesn't exist because it was generated by macro.